### ISSUE TITLE
Leshan server REST API

### ISSUE THREAD

**Issue by onovo1** [eclipse-leshan/leshan#1533]
### Question

Hi,

is there any documentation of the REST API supported by the Leshan server? I couldn't find anything and going through the code to figure out the options is a bit painful.

I found this link but it is not updated and/or well documented with proper examples:

http://robertsrhapsody.blogspot.com/2018/01/eclipse-leshan-rest-apis.html

Thanks!

**Comment by sbernard31** [comment]
> is there any documentation of the REST API supported by the Leshan server

This is REST API of leshan-server-**demo**. This API is not intended to be reused, so of course there is no documentation.

You get look at : 
 - https://github.com/eclipse-leshan/leshan/wiki/F.A.Q.#could-i-reuse-leshan--demo-
 - https://github.com/eclipse-leshan/leshan/wiki/F.A.Q.#is-there-a-rest-api-for-leshan-server-demo-

**Comment by onovo1** [comment]
Thank you for your message!

I'm aware that the REST API in the demo Leshan server is not intended for commercial use. In our case, we are utilizing it solely as a testing tool.

Our goal is to create new resources in the Leshan server using the REST API, but we have encountered challenges in accomplishing this task.

Our existing code is outlined below:

```
  type: "POST",
  url: "https://leshan.eclipseprojects.io/api/clients/IO_client/",
  payload: {"id": 1, "object": 26241, "resources": [{"kind": "singleResource", "id": 5601, "type": "FLOAT", "value": "13.1"}]}

```                

Thanks!

**Comment by sbernard31** [comment]
>  I'm aware that the REST API in the demo Leshan server is not intended for commercial use. In our case, we are utilizing it solely as a testing tool.

Yep it seems there is a lot of people who do that. :thinking: 
That sounds like a common need, so maybe one day we should try to provide a real solution about that but currently there is very few people who are regularly involved in the project and so not enough manpower to enlarge the project scope. 

Could you described a bit more how you are using it ? 

> Our goal is to create new resources in the Leshan server using the REST API, but we have encountered challenges in accomplishing this task.

Did you try the tips about : 
 - launching your web browser with developer tools
 - use web UI for feature you want to access with REST API
 - see how looks like  HTTP request / response

 ?
 
 Maybe a missing media-type ?

**Comment by onovo1** [comment]
> Could you described a bit more how you are using it ?

We have a REST client and use that client to automatically query, create, update and delete resources in the demo leshan server. Actually, it is pretty useful tool. So, we managed to update and delete resources but we can't create new resources for some reason.

> Did you try the tips about :
> 
>     launching your web browser with developer tools
>     use web UI for feature you want to access with REST API
>     see how looks like HTTP request / response
> 

we did, using the web UI works. However, it doesn't align with our primary use case. Our objective is to automate the process through our client REST API. The web UI requires manual interaction, and it is notably slower.

> 
Maybe a missing media-type ?

It is possible, but we haven't managed to figure out what is the problem and tand the error message we received from the Leshan server lacks clarity in describing the problem.

**Comment by sbernard31** [comment]
> we did, using the web UI works. However, it doesn't align with our primary use case. Our objective is to automate the process through our client REST API. The web UI requires manual interaction, and it is notably slower.

The idea is to use Web UI + web browser dev tools to explore the REST API and then you know how to use it in your code.

> the error message we received from the Leshan server lacks clarity in describing the problem.

Share the error message + share a `wget` command which reproduce the issue. 
And I could try to help.

**Comment by onovo1** [comment]
> The idea is to use Web UI + web browser dev tools to explore the REST API and then you know how to use it in your code.

Yes, you are absolutely right and this is how we have done it so far but the Web UI does not give the option of creating new resources, right? only new instances. 

So, when we try to create a new resource using the REST API, for example:

```
  type: "POST",
  url: "https://leshan.eclipseprojects.io/api/clients/IO_client/",
  payload: {"id": 1, "object": 26241, "resources": [{"kind": "singleResource", "id": 5601, "type": "FLOAT", "value": "13.1"}]}

```

The leshan server returns a 200 OK message with empty payload but the resource is not created. :/

**Comment by sbernard31** [comment]
> Yes, you are absolutely right and this is how we have done it so far but the Web UI does not give the option of creating new resources, right? only new instances.

Oh I don't get that. 

You want to create "ressource" or you wanted to say "resource instance" ? 
Because creating "ressource" doesn't make sense to me :thinking: 
And for "resource instance" this could make sense but this is not supported by LWM2M v1.1.1

See  [LWM2Mv1.1.1@core§6.3.6. Create Operation](https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-3-6-0-636-Create-Operation)

> The leshan server returns a 200 OK message with empty payload but the resource is not created. :/

That sounds strange but again give me a `wget` and I can investigate.

**Comment by sbernard31** [comment]
> I meant "enabling a new object" in the client.

AFAIK, this is not allowed by the LWM2M specification too.

**Comment by onovo1** [comment]
that is related with the CREATE operation, right?

[LWM2Mv1.1.1@core§6.3.6. Create Operation](https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-3-6-0-636-Create-Operation)

**Comment by sbernard31** [comment]
> that is related with the CREATE operation, right?

Yep, but more generally this has not so much sense to add support of an object just by sending a request.
I mean generally adding support of an object means "implement some specific behavior" at client side. 

> The leshan server returns a 200 OK message with empty payload

I was able to reproduce and even if REST API is not intended to be reused current behavior is not so good.  I will adapt the code to return a BAD_REQUEST in that case.

**Comment by sbernard31** [comment]
> The leshan server returns a 200 OK message with empty payload

Ok I fixed it with 73916b77150e151d8519b0d1f39bc8d91e8835a7
This is already in `master` (will be soon available in sandbox) and will be in next 2.0.0-M14 release

**Comment by sbernard31** [comment]
@onovo1 let me know if we should close this issue ?

**Comment by onovo1** [comment]
>Ok I fixed it with https://github.com/eclipse-leshan/leshan/commit/73916b77150e151d8519b0d1f39bc8d91e8835a7
>This is already in master (will be soon available in sandbox) and will be in next 2.0.0-M14 release

thanks for such a quick fix!

> I mean generally adding support of an object means "implement some specific behavior" at client side.

OK, I understand. So, the current behavior allowed on a server should be to request information and update the already existing objects, right?

I think we also got a 200 OK when trying to delete an objects from the server using the REST API, but such an object was never deleted.

**Comment by sbernard31** [comment]
> OK, I understand. So, the current behavior allowed on a server should be to request information and update the already existing objects, right?

Yep

> I think we also got a 200 OK when trying to delete an objects from the server using the REST API, but such an object was never deleted.

OK thx for reporting this :pray: 
I will try to look at it (today or next week)

**Comment by sbernard31** [comment]
> I think we also got a 200 OK when trying to delete an objects from the server using the REST API, but such an object was never deleted.

I tried that and get Bad Request
```
>> curl -i  -X DELETE  "http://localhost:8088/api/clients/mooky/3033"
HTTP/1.1 400 Bad Request
X-Powered-By: Express
connection: close
date: Fri, 10 Nov 2023 14:37:48 GMT
server: Jetty(9.4.51.v20230217)
Transfer-Encoding: chunked
```

Could you provide a way to reproduce ? :pray:

**Comment by onovo1** [comment]
Actually, it is deleting the instance in the server and the client:

```
curl -X DELETE "https://leshan.eclipseprojects.io/api/clients/IO_client/3442/0"
{"status":"DELETED(202)","valid":true,"success":true,"failure":false}

curl -X DELETE "https://leshan.eclipseprojects.io/api/clients/IO_client/3303/0"
{"status":"DELETED(202)","valid":true,"success":true,"failure":false}
```

Is that the right behavior? 

Now, I remember, we expected the same behavior than DELETE for enabling an object instance in the server using the REST API.

**Comment by sbernard31** [comment]
> Is that the right behavior?

Yep,
 -  Delete is about deleting Object Instance. (See : [LWM2Mv1.1.1@core§6.3.6. Create Operation](https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-3-6-0-636-Create-Operation))
- Create is about Creating Object Instance. (See : [LWM2Mv1.1.1@core§6.3.7. Delete Operation](https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-3-7-0-637-Delete-Operation))

There is no way to enable/disable object at server initiative.

**Comment by onovo1** [comment]
> There is no way to enable/disable the whole object at server initiative

Is there a way to create a new instance of an enabled object using the REST API, then?

**Comment by sbernard31** [comment]
Yep used by `[Create]` button in Web UI.

**Comment by onovo1** [comment]
> Yep used by [Create] button in Web UI.

I meant, using the REST API. I was just wondering if currently that was possible.

**Comment by sbernard31** [comment]
Yeah it is possible using REST API.
Just click on `[Create]` button in Web UI, and use your browser devtools and you will see how to use the REST API. 

Please, let me know is there is something unclear because I repeat myself several time on that point and I don't know how to explain it better :grimacing:

**Comment by onovo1** [comment]
> Please, let me know is there is something unclear because I repeat myself several time on that point and I don't know how to explain it better 😬

Everything is clear now. :)

**Comment by sbernard31** [comment]
You we close this issue ?

**Comment by onovo1** [comment]
yes, please.

It would also be beneficial to have a proper documentation for the REST API. I understand that this may not be the highest priority given other important features that need attention, but having the documentation for the REST API would be a valuable addition, and it's unfortunate that the API is currently hidden.

**Comment by sbernard31** [comment]
> having the documentation for the REST API would be a valuable addition, and it's unfortunate that the API is currently hidden.

Yep I totally understand you from a user point of view.
But as I explained, when we created our  REST API, we didn't have in mind to make it reusable.
This API was just made to be used by the front-end UI of our server-demo.

So not so much effort was invest in it. Code quality of the demo is much lower than the library one.
And for now we haven't time enough to make it a "real  component" with its own documentation. 

Even if it seems that there is really a community need behind this :
 - https://github.com/eclipse/wakaama/issues/617
 - https://github.com/zephyrproject-rtos/zephyr/pull/64013
